pkg/cmd/cli/debug/debug.go

```diff
@@ -0,0 +1,140 @@
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	crashdCmd "github.com/vmware-tanzu/crash-diagnostics/cmd"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/vmware-tanzu/velero/pkg/client"
+	"github.com/vmware-tanzu/velero/pkg/cmd"
+)
+
+//go:embed cshd-scripts/velero.cshd
+var scriptBytes []byte
+
+type option struct {
+	// workdir for crashd will be $baseDir/velero-debug
+	baseDir string
+	// the namespace where velero server is installed
+	namespace string
+	// the absolute path for the log bundle to be generated
+	outputPath string
+	// the absolute path for the kubeconfig file that will be read by crashd for calling K8S API
+	kubeconfigPath string
+	// optional, the name of the backup resource whose log will be packaged into the debug bundle
+	backup string
+	// optional, the name of the restore resource whose log will be packaged into the debug bundle
+	restore string
+}
+
+func (o *option) bindFlags(flags *pflag.FlagSet) {
+	flags.StringVar(&o.outputPath, "output", "", "The path of the bundle tarball, by default it's $HOME/bundle.tar.gz. Optional")
+	flags.StringVar(&o.backup, "backup", "", "The name of the backup resource whose log will be collected, no backup logs will be collected if it's not set. Optional")
+	flags.StringVar(&o.restore, "restore", "", "The name of the restore resource whose log will be collected, no restore logs will be collected if it's not set. Optional")
+}
+
+func (o *option) asCrashdArgs() string {
+	return fmt.Sprintf("output=%s,namespace=%s,basedir=%s,backup=%s,restore=%s,kubeconfig=%s",
+		o.outputPath, o.namespace, o.baseDir, o.backup, o.restore, o.kubeconfigPath)
+}
+
+func (o *option) complete(f client.Factory, fs *pflag.FlagSet) error {
+	if len(o.outputPath) > 0 {
+		absOutputPath, err := filepath.Abs(o.outputPath)
+		if err != nil {
+			return fmt.Errorf("invalid output path: %v", err)
+		}
+		o.outputPath = absOutputPath
+	}
+	tmpDir, err := ioutil.TempDir("", "crashd")
+	if err != nil {
+		return err
+	}
+	o.baseDir = tmpDir
+	o.namespace = f.Namespace()
+	kp := kubeconfig(fs)
+	o.kubeconfigPath, err = filepath.Abs(kp)
+	if err != nil {
+		return fmt.Errorf("invalid kubeconfig path: %s, %v", kp, err)
+	}
+	return nil
+}
+
+// NewCommand creates a cobra command.
+func NewCommand(f client.Factory) *cobra.Command {
+	o := &option{}
+	c := &cobra.Command{
+		Use:   "debug",
+		Short: "Generate debug bundle",
+		Long: `Generate a tarball containing the logs of velero deployment, plugin logs, restic DaemonSet, 
+specs of BackupStorageLocations, PodVolumeBackups, PodVolumeRestores, and optionally the specs and logs of backup and restore.`,
+		Run: func(c *cobra.Command, args []string) {
+			defer func(opt *option) {
+				if len(o.baseDir) > 0 {
+					if err := os.RemoveAll(o.baseDir); err != nil {
+						fmt.Fprintf(os.Stderr, "Failed to remove temp dir: %s: %v\n", o.baseDir, err)
+					}
+				}
+			}(o)
+			flags := c.Flags()
+			err := o.complete(f, flags)
```

Comment:
Can we do some validation of the `backup` and `restore` arguments here before they are passed to crashd usin something like this: https://github.com/vmware-tanzu/velero/blob/main/pkg/cmd/cli/backup/get.go#L53-L54)?

I tried this out locally with a backup that didn't exist and the script failed when calling `backup logs`. Unfortunately crashd didn't give the stderr to say that the backup didn't exist. I tried looking at the crashd functions that we can call, and it doesn't look like it supports providing the stderr. I think it would be easier to just validate the existence of those objects here before calling crashd.

Reply:
I made the update to add some validation against the kubeconfig, kubecontext, namespace and the backup and restore.